x-pack/plugins/fleet/server/services/security/uninstall_token_service/index.ts

```diff
@@ -202,19 +202,31 @@ export class UninstallTokenService implements UninstallTokenServiceInterface {
   }
 
   private async getDecryptedTokensForPolicyIds(policyIds: string[]): Promise<UninstallToken[]> {
+    /** If BATCH_SIZE is too large, we get an error of `too_many_nested_clauses`.
+     *  Assuming that `max_clause_count` >= 1024, and experiencing that batch size should be less than half
+     *  than `max_clause_count` with our current query, batch size below 512 should be okay on every env.
+     */
+    const BATCH_SIZE = 500;
+
     const tokenObjectHits = await this.getTokenObjectsByIncludeFilter(policyIds);
 
     if (tokenObjectHits.length === 0) {
       return [];
     }
 
-    const filter: string = tokenObjectHits
-      .map(({ _id }) => {
-        return `${UNINSTALL_TOKENS_SAVED_OBJECT_TYPE}.id: "${_id}"`;
-      })
-      .join(' or ');
+    const filterEntries: string[] = tokenObjectHits.map(
+      ({ _id }) => `${UNINSTALL_TOKENS_SAVED_OBJECT_TYPE}.id: "${_id}"`
+    );
```

Comment:
If you want to avoid batching, building an Elasticsearch DSL object directly can help. The SO find method accepts both Elasticsearch DSL and KQL. The [`terms` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-terms-query.html) can take an array of values to compare against a single field and only counts as a single clause, so something like 
```
const filter = {
  terms: {
    [`${UNINSTALL_TOKENS_SAVED_OBJECT_TYPE}.id`]: tokenObjectHits.map(({ _id }) => _id),
  }
};
```
could put up to 65536 _id values in a single clause and avoid the max clause count issues.

We've run into max clause count a number of times on the Detections and Response team - I saw it mentioned on the blocker issue so I thought I'd share in case it's useful info here.

Reply:
We use it [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/lists/server/services/exception_lists/build_exception_filter.ts#L539) for example

Reply:
Have you tried a KQL query that will produce a single clause, I think a format like `${UNINSTALL_TOKENS_SAVED_OBJECT_TYPE}.id: (1 or 2 or 3)` may work.

Reply:
just tried it, depending on quote usage it's either an unhandled error somewhere in `validateConvertFilterToKueryNode()`, or a handled error like `The key is empty and needs to be wrapped by a saved object type like...`

Reply:
Apologies, I got confused by the `KueryNode` type and thought that was pure ES DSL. I see some context in a [slack discussion](https://elastic.slack.com/archives/C5TQ33ND8/p1704828296830649) about why the SO client doesn't support DSL.

I also don't see a quick way to inject the `terms` clause into the `KueryNode`. Sorry for the distraction!

Reply:
no worries at all, thanks a lot for the input! and thanks for the slack link, haven't seen that yet